### DISCUSSION TITLE
Shape of probabilities generated by predict_survival_function

### DISCUSSION THREAD

**Discussion by shaurya-44** [sebp/scikit-survival#451] [Q&A] [answered]
Hi,

Thanks for your comprehensive survival analysis implementation, which has been very helpful with my current people analytics use case, I had a query that has risen due to a recent constraint-based change.

Due to both the nature of the end user input data and local memory constraints, the encoding being performed had to be modified, interestingly leading to a decrease in the "time point" columns/steps (that I had interpreted to be consistent to indicate number of days) generated by the predict_survival_function for the Random Survival Forest, Coxnet, and Gradient Boosting models, almost halving the available time points for which employee probabilities could be fetched, would appreciate gaining some insight into how this could possibly be prevented or what could be causing this decrease, as I may have misunderstood the significance of the generated survival probabilities.

Thank you

**Comment by sebp** [comment]
I'm not sure I fully understood your question, so please correct me if I'm wrong.

If you are referring to which time points the survival functions produced by `predict_survival_function` are evaluated at, the answer is the estimator's `unique_times_` attribute, which is the unique time points of the training data where events occurred. If the number of unique time points is halved, it would suggest you have more tied times than before.

**Reply by shaurya-44** [reply]
I see, thank you for your reply @sebp, apologies for my phrasing.

Running `cn_pred.unique_times_` confirms that the time points have indeed decreased to 2970 from 4000+.

I'd be very grateful for some clarity regarding: 

1) why are the time points for the new subjects (for whom the fitted model is now generating fresh probability predictions) the same as those that the models were fitted on, (if this could be manually adjusted during or before training when the model objects are created?), and 

2) do you think my base working assumption of these models' results is incorrect - where I would get the predictions with the set parameter `return_array=True`, then slice the ndarray's at a future time point -> 

`new subject's elapsed span + desired prediction time point` 

> (all accounted for in days, right from preprocessing before till interpretation of the resulting probabilities)

**Reply by sebp** [reply, marked as answer]
> why are the time points for the new subjects (for whom the fitted model is now generating fresh probability predictions) the same as those that the models were fitted on, (if this could be manually adjusted during or before training when the model objects are created?), and

Without going into the details, the reason is that the models do not make an assumption about the exact form of the survival function, thus they can only predict for those time points where an event was observed in the training data. Theoretically, it is possible to predict beyond those time point by making assumptions on the from of the survival function (e.g. following a Weibull distribution). However, scikit-survival currently does not contain such models.

> do you think my base working assumption of these models' results is incorrect - where I would get the predictions with the set parameter return_array=True, then slice the ndarray's at a future time point

I'm afraid that's not possible, because the last entry in the array corresponds to `unique_times_[-1]`. You cannot slice it in a way to predict probabilities beyond that, unless the probability at `unique_times_[-1]` is already zero.
